docs/design/git-submodule-storage.md

```diff
@@ -0,0 +1,138 @@
+is stored as a local Git repository.
+
+Rolling our own Git fetch is too complex to be worth the effort.
+
+### "jj op restore" and operation log format
+
+We want `jj op restore` in the superproject to restore to an "expected" state in
+the submodule, though the definition of "expected" is up for debate. At the very
+least, `jj op restore` should restore the submodule working copy, but we may
+want to support the full list of operations at some point.
+
+Currently, the operation log only references objects and refs in the
+superproject, so it is likely that proposed approaches will need to extend this
+format. It is also worth considering that submodules may be added, updated or
+removed in superproject commits, thus the list of submodules is likely to change
+over the repository's lifetime.
+
+### Nested submodules
+
+Git submodules may contain submodules themselves, so our chosen storage schemes
+should support that.
+
+### Bonus: supporting future extensions
+
+There are certain extension we may want to make in the future, but are non-goals
+today. When evaluating approaches, consider these as low-priority "bonus"
+points; they shouldn't weight heavily in our decision-making, but all else being
+equal, we should prefer the approach that is more easily extended.
+
+These extensions are:
+
+- Non-git subrepos
+- Colocated Git repos
+
+## Possible approaches
+
+### Approach 1: Git repos in the main Git backend
+
+Since the Git backend contains a Git repository, an 'obvious' default would be
+to store them in the Git superproject the same way Git does, i.e. in
+`.git/modules`. This should allow us to _mostly_ reuse the existing backend
+interfaces. However, it is not likely that all backends will support Git
+submodules (see "Approaches Rejected"), and we will have to adjust the backend
+interfaces accordingly.
+
+TODO(chooglen): Discuss operation log
+TODO(chooglen): Discuss nested submodules
+
+TODO(chooglen) I'm not sure the backend interfaces mean that the rest of jj can
+  just 'ignore' the existence of Git submodules, e.g. when populating the
+  working copy, can we just ask the Git backend to give the tree structure we
+  need without knowing about submodules?
+
+### Approach 2: Store Git submodules as full jj repos
+
+This would be somewhere in `.jj` but outside of `.jj/store`. We would then
+expose a "submodules" interface that gets hooked up to the relevant machinery
+(e.g. updating the working copy).
+
+TODO(chooglen): Discuss operation log
+TODO(chooglen): Discuss nested submodules
+
+### Approach 3: Store Git submodules as alternate Git backends
+
+First, teach jj to use multiple backends in a way that works for Git submodules.
+A scheme could be based on a generalization of how Git submodules work: we could
+say that the "main" backend populates the root of the working copy. During this
+process, it can leave some subdirectories unpopulated and include a hint to jj
+to use a different backend (i.e. the submodule backend) instead.
```

Comment:
> How do you imagine cross-repo coordination and recursive fetches?

What sorts of cross-repo coordination do you have in mind, btw?

Right, this paragraph only describes how coordination would happen when populating the working copy.
For fetching, my idea (which I haven't formally proposed yet) is to ignore the working copy, and instead, iterate through whatever submodules have been set up (i.e. iterate through the submodules backends). This is actually different from git/git, which primarily works on submodules in the working copy, but this doesn't really work well in general. E.g if you only consider the working tree, `main` and `topic` have two different sets of submodules, and you `git checkout main && git fetch --recurse-submodules`, you won't end up fetching the submodules for topic. This _exact_ scenario won't occur any more because I patched it a while ago, but that experience left me confident that there are many more of these problems lurking under the surface.

My theory is that this can use a similar approach whenever we "recurse through submodules", but that will have to be its own discussion though.

> Isn't this approach also very complex?

Yes, it requires quite a lot of coordination between the 'main' backend and the submodule backends.

Reply:
> What sorts of cross-repo coordination do you have in mind, btw?

Nothing special, just usual cd'ing into submodule A doing some work there, going back to the superproject continuing the work there and then going into submodules B and C for some work.  

I imagine that this will generate quite a bit of work for all Jujutsu backends. I think this chatter can be quite hard to capture correctly. But this is not a strong opinion at all, so change my mind :-).

> Right, this paragraph only describes how coordination would happen when populating the working copy.
For fetching, my idea (which I haven't formally proposed yet) is to ignore the working copy, and instead, iterate through whatever submodules have been set up (i.e. iterate through the submodules backends).

Sounds interesting.

>  This is actually different from git/git, which primarily works on submodules in the working copy, but this doesn't really work well in general. E.g if you only consider the working tree, main and topic have two different sets of submodules, and you git checkout main && git fetch --recurse-submodules, you won't end up fetching the submodules for topic. This exact scenario won't occur any more because I patched it a while ago, but that experience left me confident that there are many more of these problems lurking under the surface.

IIUC and please correct me if I'm wrong, previously `git checkout main && git fetch --recursive submodules` only fetched the submodules for `main` instead of  `main | topic` although `topic` is in the same subtree. This behavior seems to be vastly underspecified, so I'd expect alot of unknown unknowns with bad interacting edge cases :-/.

Lets hope we can get a simpler model for Jujutsu. 

> My theory is that this can use a similar approach whenever we "recurse through submodules", but that will have to be its own discussion though.

Sounds great.   

> Yes, it requires quite a lot of coordination between the 'main' backend and the submodule backends.

If your positive that this doesn't blow the maintenance and complexity budget, I'll be happy to accept it.

Reply:
> git checkout main && git fetch --recursive submodules only fetched the submodules for main instead of main | topic although topic is in the same subtree. 

I'm not sure what you mean by subtree in this context, but I'll give an example and you can see if that lines up with what you're thinking :)

In their 'absorbed' form, submodules live in `.git/modules/`, but (until recently) a submodule only performs a recursive fetch if it is _also_ in the working tree. So, e.g. if you `git checkout --recurse-submodules main`, resulting in:

```
.
├── .git/
│   └── modules/
│       ├── main-submodule
│       └── topic-submodule
├── foo
├── bar
└── main-submodule/
```

and you `git fetch --recurse-submodules`, you wouldn't touch `topic-submodule`. My patch makes submodules in `.git/modules` fetch too, regardless of whether they're in the working tree.

> This behavior seems to be vastly underspecified, so I'd expect alot of unknown unknowns with bad interacting edge cases :-/.

Yup :) Git's submodule support is ad-hoc and dependent on the command in question. The behavior is also often underspecified even for a single command.

> If your positive that this doesn't blow the maintenance and complexity budget, I'll be happy to accept it.

I suspect that it won't be more conceptually complicated than any of the alternatives here; they will all have to find a way to coordinate between superproject and submodule anyway, so whether or not this becomes hard to maintain will probably come down to execution. That remains to be seen, but I think it's feasible, or at the very least, I'll say something if something unexpected comes up during implementation.